### ISSUE TITLE
Memory leak in keras.backend preventing graphs from being GC'ed

### ISSUE THREAD

**Issue by msg555** [tensorflow/tensorflow#42022]
In some situations (i.e. [when outside of an eager execution context](https://github.com/tensorflow/tensorflow/blob/v2.3.0/tensorflow/python/keras/backend.py#L394)) a Tensor value will be inserted into [_GRAPH_LEARNING_PHASES](https://github.com/tensorflow/tensorflow/blob/v2.3.0/tensorflow/python/keras/backend.py#L6389). This is a problem because the Tensor value itself references the graph which is suppose to be weakly held in memory by this dictionary. With the implementation of weakref.WeakKeyDictionary a value that strongly references its key will forever prevent the key from being reclaimed. This means the entire graph is permanently stuck in memory unless the entry is explicitly deleted.

This issue is present in tf 2.3.0

Simple repro code

```
import gc

import psutil
import tensorflow as tf

FIX_LEAK = False

def _get_rss():
    return psutil.Process().memory_info().rss

def _test_iteration():
    with tf.Graph().as_default() as graph:
        inputs = tf.keras.Input(shape=(8,), dtype="int32")
        embedding_layer = tf.keras.layers.Embedding(
            input_dim=128,
            output_dim=128,
            input_length=8,
        )(inputs)
        tf.keras.layers.LSTM(
            256,
            dropout=0.5,
        )(embedding_layer)
        
    if FIX_LEAK:
        import tensorflow.python.keras.backend as backend
        backend._GRAPH_LEARNING_PHASES.pop(graph, None)
    
for _ in range(128):
    _test_iteration()

    gc.collect()
    print("RSS", _get_rss())
```

With FIX_LEAK as False the memory usage grows forever (about 8MB or so for me each iteration). If FIX_LEAK is True the memory stays stable. I'm not super familiar with TF/keras, there may be a more bare bones example. In particular it seems that the dropout parameter was needed to trigger the learning_phrase Tensor to be created.

**Comment by jammm** [comment]
This could be the cause for the leak in this issue as well https://github.com/tensorflow/tensorflow/issues/41718

I enabled eager execution via `tf.config.experimental_run_functions_eagerly(True)` and the memory usage seems quite stable around 1GB, compared to ~7GB+ that I was getting without eager mode.

**Comment by amahendrakar** [comment]
@msg555,
Could you please provide a minimal working example to reproduce the issue reported here. Thanks!

**Comment by msg555** [comment]
> @msg555,
> Could you please provide a minimal working example to reproduce the issue reported here. Thanks!

Added an example

**Comment by amahendrakar** [comment]
@msg555,
Thank you for the update. 

I was able to reproduce the issue with TF v2.3, please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/6c18a1fd9f1a94c7fe5e942118239bee/42022.ipynb). Thanks!

**Comment by msg555** [comment]
> This could be the cause for the leak in this issue as well #41718
> 
> I enabled eager execution via `tf.config.experimental_run_functions_eagerly(True)` and the memory usage seems quite stable around 1GB, compared to ~7GB+ that I was getting without eager mode. FWIW, I noticed that CPU usage maxes out at 1 core only, compared to multiple cores when using graph mode.
> 
> The example code in that issue could be used to reproduce and check this behavior.

@jammm You can test if it's the same issue by examining `tensorflow.python.keras.backend._GRAPH_LEARNING_PHASES`. If it continues to grow over time with more elements than it's likely the same issue.

**Comment by tomerk** [comment]
Sorry for the delayed response here! The backend graph (& backend learning phases) have been historically prone to memory leaks, and we have put a lot of focus into getting rid of the global backend graph entirely in 2.4 to fix these leaks.

That said, the specific code snippet in this github issue enters a v1 `tf.Graph.as_default()` explicitly which is a legacy v1 graph compatibility mode. It will not benefit from these improvements, and will not receive performance improvements in the future. If you use eager execution, with tf.function to create/enter graphs as needed, the specific memory leak reported here will not be an issue in 2.4.

As such, closing this issue but happy to answer any follow-on questions.

**Comment by jammm** [comment]
@tomerk  Thanks a lot for your reply! I'm currently using tf.function in many places. How do I make sure that it's running in eager mode in code outside of tf.function?

**Comment by tomerk** [comment]
If you're using tf2 and haven't explicitly used the `tf.compat.v1.disable_eager_execution()` command it should be eager. If you're not sure and you want to check it programmatically, you can use `tf.executing_eagerly()`: https://www.tensorflow.org/api_docs/python/tf/executing_eagerly

It will return True when running eagerly, and False when tracing a tf.function into a graph.

**Comment by tomerk** [comment]
I'm not sure if your issue #41718 is related to this though, that one will necessitate a closer look

**Comment by tomerk** [comment]
Hi @saikumarchalla is there a specific reason you re-opened this issue?
